### PR TITLE
whitebit - transfer

### PR DESCRIPTION
**Pull request by pcriadoperez**

### CODE REVIEW

**Comment by pcriadoperez** [comment]
When testing with python and php, it fails when calling `currencyToPrecision` because the precision of the currencies are undefined.
The two possible solutions I can think of is:
1) Allow in php and python for precision to be undefined when calling `currencyToPrecision`
2) For whitebit the precision is returned when calling fetchMarkets, we can use that information to infer the precision when calling `fetchCurrencies`

File: js/whitebit.js [hunk 1]
```diff
@@ -1151,6 +1161,42 @@ module.exports = class whitebit extends Exchange {
...
+            'info': response,
+            'status': 'pending',
+        };
+        if (fillTransferResponseFromRequest) {
+            transfer['currency'] = code;
+            transfer['amount'] = amount;
+            transfer['fromAccount'] = fromAccount;
+            transfer['toAccount'] = toAccount;
+        }
+        return transfer;
```
Comment (kroitor) [inline js/whitebit.js:1197]: Let's move lines 1181-1197 to `parseTransfer`.

Comment (pcriadoperez) [reply js/whitebit.js:1197]: I agree it's good to have it all in parseTransfer, however so far I haven't been doing it not to change the signature of parseTransfer. What's the best way to pass in amount, fromAccount and toAccount? Should we add like a params argument to parseTransfer? Or add those fields as extra parameters to parseTransfer?

Comment (kroitor) [reply js/whitebit.js:1197]: Let's make sure that the returned transfer structure is fully unified, including all fields: https://github.com/ccxt/ccxt/wiki/Manual#transfer-structure

Comment (kroitor) [reply js/whitebit.js:1197]: The point of parseTransfer is to make sure that we return the same unified structure format, so, we can do this (pseudocode):

```JavaScript
parseTransfer (transfer, currency = undefined) { 
    //
    //    []
    //
    return {
        'info': transfer,
        'timestamp': undefined,
        'datetime': undefined,
        // ... other fields of the unified transfer structure ...
    }
}

async transfer (code, amount, ...) {
    await exchange.loadMarkets ();
    const currency = this.currency (code);
    // ...
    const response = await this.privateXXXXXX (...);
    //
    //    []
    //
    const transfer = this.parseTransfer (transfer, currency);
    if (fill...) {
        transfer['toAccount'] = toAccount;
        transfer['fromAccount'] = fromAccount;
    }
    return transfer;
}
```

File: js/whitebit.js [hunk 2]
```diff
@@ -1151,6 +1161,55 @@ module.exports = class whitebit extends Exchange {
...
 
+    async transfer (code, amount, fromAccount, toAccount, params = {}) {
+        await this.loadMarkets ();
+        const currency = this.currency (code);
+        let type = undefined;
+        if (fromAccount === 'main' && toAccount === 'spot') {
+            type = 'deposit';
+        } else if (fromAccount === 'spot' && toAccount === 'main') {
+            type = 'withdraw';
+        }
```
Comment (kroitor) [inline js/whitebit.js:1172]: Let's support both the unified types (`main`, `spot`) and the exchange-specific types (`main`, `trade`) here:

```JavaScript
        const accountsByType = this.safeValue (this.options, 'accountsByType');
        const fromAccountId = this.safeString (accountsByType, fromAccount, fromAccount);
        const toAccountId = this.safeString (accountsByType, toAccount, toAccount);
        let type = undefined;
        if (fromAccount === 'main' && toAccountId === 'trade') {
            type = 'deposit';
        } else if (fromAccount === 'trade' && toAccount === 'main') {
            type = 'withdraw';
        }
        if (type === undefined) {
            throw new ExchangeError (this.id + ' transfer() only allows transfers between main account and spot (trade) account');
        }
```

Also, let's use uniform exception messages (everywhere):

```JavaScript
throw new BadSymbol (this.id + ' fetchFundingRate() supports contracts only');
throw new ArgumentsRequired (this.id + ' fetchOpenOrders() requires a symbol argument');
...
throw new ExceptionClassName (this.id + ' methodName() blablabla');
```

**Pull request merged**